enhancements/sig-architecture/136-placement-cel-selector/README.md

```diff
@@ -0,0 +1,343 @@
+	// LabelSelector represents a selector of ManagedClusters by label
+	// +optional
+	LabelSelector metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// ClaimSelector represents a selector of ManagedClusters by clusterClaims in status
+	// +optional
+	ClaimSelector ClusterClaimSelector `json:"claimSelector,omitempty"`
+
+	// CelSelector represents a selector of ManagedClusters by CEL expressions on ManagedCluster fields
+	// +optional
+	CelSelector ClusterCelSelector `json:"celSelector,omitempty"`
+}
+
+// ClusterCelSelector is a list of CEL expressions. The expressions are ANDed.
+type ClusterCelSelector struct {
+	// +optional
+	CelExpressions []string `json:"celExpressions,omitempty"`
+}
+```
+
+### ManagedCluster CEL Library
+
+A ManagedClusterLib will be added.
+- Variable `managedCluster` will be added.
+- Function `scores` will be added. 
+- Function `versionIsGreaterThan`, `versionIsLessThan` will be added. 
+- Function `quantityIsGreaterThan`, `quantityIsLessThan` will be added. 
+
+```go
+// ManagedClusterLib defines the CEL library for ManagedCluster evaluation.
+// It provides functions and variables for evaluating ManagedCluster properties
+// and their associated resources.
+//
+// Variables:
+//
+// managedCluster
+//
+// Provides access to ManagedCluster properties.
+//
+// Functions:
+//
+// scores
+//
+// Returns a list of AddOnPlacementScoreItem for a given cluster and AddOnPlacementScore resource name.
+//
+//	scores(<ManagedCluster>, <string>) <list>
+//
+// The returned list contains maps with the following structure:
+//   - name: string - The name of the score
+//   - value: int - The numeric score value
+//   - quantity: number|string - The quantity value, represented as:
+//     * number: for pure decimal values (e.g., 3)
+//     * string: for values with units or decimal places (e.g., "300Mi", "1.5Gi")
+//
+// Examples:
+//
+//	managedCluster.scores("cpu-memory") // returns [{name: "cpu", value: 3, quantity: 3}, {name: "memory", value: 4, quantity: "300Mi"}]
+//
+// Version Comparisons:
+//
+// versionIsGreaterThan
+//
+// Returns true if the first version string is greater than the second version string.
+// The version must follow Semantic Versioning specification (http://semver.org/).
+// It can be with or without 'v' prefix (eg, "1.14.3" or "v1.14.3").
+//
+//	versionIsGreaterThan(<string>, <string>) <bool>
+//
+// Examples:
+//
+//	versionIsGreaterThan("1.25.0", "1.24.0") // returns true
+//	versionIsGreaterThan("1.24.0", "1.25.0") // returns false
+//
+// versionIsLessThan
+//
+// Returns true if the first version string is less than the second version string.
+// The version must follow Semantic Versioning specification (http://semver.org/).
+// It can be with or without 'v' prefix (eg, "1.14.3" or "v1.14.3").
+//
+//	versionIsLessThan(<string>, <string>) <bool>
+//
+// Examples:
+//
+//	versionIsLessThan("1.24.0", "1.25.0") // returns true
+//	versionIsLessThan("1.25.0", "1.24.0") // returns false
```

Comment:
Can the k8s semver CEL library be used to handle these use cases? https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/semverlib.go

Reply:
Yes, let me try.

Reply:
Hello @jpbetz it seems the Semver version with a v prefix is not supported in the library? See [this case](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/semver_test.go#L135).  What does ["tolerant parsing"](https://github.com/kubernetes/kubernetes/blob/v1.32.0/staging/src/k8s.io/apiserver/pkg/cel/library/semverlib.go#L55 ) mean and how to support v as a prefix?

Reply:
Looks like the underlying library has a [ParseTolerant](https://github.com/kubernetes/kubernetes/blob/fe7af1c68bc0c772e78d57b5c38ac4ae8ef122ab/vendor/github.com/blang/semver/v4/semver.go#L240) function that: "currently trims spaces, removes a "v" prefix, adds a 0 patch number to versions with only major and minor components specified, and removes leading 0s."

I'm not sure why it is alluded in the semver CEL library since it doesn't call into it.

I've opened https://github.com/kubernetes/kubernetes/pull/130648 to add this support directly to the library.

As a short term mitigation something like`"v1.0.0".find("[^v].*")` should work.